hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -479,6 +510,200 @@ public Hashtable<String, String> getPathStatus(final Path path,
     }
   }
 
+  /**
+   * Orchestrates the copying of blob from given source to a given destination.
+   * @param srcPath source path
+   * @param dstPath destination path
+   * @param tracingContext object of TracingContext used for the tracing of the
+   * server calls.
+   * @throws AzureBlobFileSystemException exception thrown from the server calls,
+   * or if it is discovered that the copying is failed or aborted.
+   */
+  @org.apache.hadoop.classification.VisibleForTesting
+  void copyBlob(Path srcPath,
+      Path dstPath,
+      TracingContext tracingContext) throws AzureBlobFileSystemException {
+    AbfsRestOperation copyOp = null;
+    try {
+      copyOp = client.copyBlob(srcPath, dstPath,
+          tracingContext);
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        final BlobProperty dstBlobProperty = getBlobProperty(dstPath,
+            tracingContext);
+        try {
+          if (dstBlobProperty.getCopySourceUrl() != null &&
+              ("/" + client.getFileSystem() + srcPath.toUri().getPath()).equals(
+                  new URL(dstBlobProperty.getCopySourceUrl()).toURI()
+                      .getPath())) {
+            return;
+          }
+        } catch (URISyntaxException | MalformedURLException e) {
+          throw new RuntimeException(e);
+        }
+      }
+      throw ex;
+    }
+    final String progress = getCopyBlobProgress(copyOp);
+    if (COPY_STATUS_SUCCESS.equals(progress)) {
+      return;
+    }
+    final String copyId = copyOp.getResult().getResponseHeader(X_MS_COPY_ID);
+    while (true) {//TODO: better condition.
```

Comment:
Is this the final condition for this progress check ?

Reply:
yes, this is final condition. removing comment.

Reply:
the condition has been made like below as suggested in other comment:
```
while (handleCopyInProgress(dstPath, tracingContext, copyId)
        == BlobCopyProgress.PENDING) {
      try {
        Thread.sleep(pollWait);
      } catch (Exception e) {

      }
    }
```